Doc/tools/extensions/grammar_snippet.py

```diff
@@ -41,98 +41,111 @@ class GrammarSnippetBase(SphinxDirective):
 
     # The option/argument handling is left to the individual classes.
 
+    grammar_re: Final = re.compile(
+        r"""
+            (?P<rule_name>^[a-zA-Z0-9_]+)     # identifier at start of line
+            (?=:)                             # ... followed by a colon
+        |
+            (?P<rule_ref>`[^\s`]+`)           # identifier in backquotes
+        |
+            (?P<single_quoted>'[^']*')        # string in 'quotes'
+        |
+            (?P<double_quoted>"[^"]*")        # string in "quotes"
+        """,
+        re.VERBOSE,
+    )
+
     def make_grammar_snippet(
         self, options: dict[str, Any], content: Sequence[str]
-    ) -> list[nodes.paragraph]:
+    ) -> list[addnodes.productionlist]:
         """Create a literal block from options & content."""
 
         group_name = options['group']
-
-        # Docutils elements have a `rawsource` attribute that is supposed to be
-        # set to the original ReST source.
-        # Sphinx does the following with it:
-        # - if it's empty, set it to `self.astext()`
-        # - if it matches `self.astext()` when generating the output,
-        #   apply syntax highlighting (which is based on the plain-text content
-        #   and thus discards internal formatting, like references).
-        # To get around this, we set it to this non-empty string:
-        rawsource = 'You should not see this.'
-
-        literal = nodes.literal_block(
-            rawsource,
+        node_location = self.get_location()
+        production_nodes = []
+        source_line = None
+        for source_line in content:
+            # Start a new production_node if there's text in the first
+            # column of the line. (That means a new rule is starting.)
+            if source_line is None or not source_line[:1].isspace():
+                production_node = addnodes.production(source_line)
```

Comment:
Here, `source_line` is passed as the first argument; I'm not sure how to update `production_node` to add the subsequent lines.

Reply:
This is just for the raw source, you can modify via `node.rawsource`, or pass a fully constructed raw source for the production.